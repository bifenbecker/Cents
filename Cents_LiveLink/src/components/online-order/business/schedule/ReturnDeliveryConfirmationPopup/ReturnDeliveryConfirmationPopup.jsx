import {useEffect, useState} from "react";
import {useSelector} from "react-redux";
import {Button, Typography, IconButton, CircularProgress} from "@material-ui/core";
import {Close} from "@material-ui/icons";
import {
  getNearStoresData,
  getReturnWindowsState,
  getScheduleDetails,
} from "components/online-order/redux/selectors";
import {useAppDispatch} from "app/hooks";
import {onlineOrderActions, onlineOrderThunks} from "components/online-order/redux";
import VanImg from "assets/images/grayscale/Van.svg";
import TurnAroundImg from "assets/images/grayscale/Turnaround Time.svg";
import {ORDER_MESSAGES, SCHEDULE, TEXT_ME_WHEN_READY, VIEWS} from "constants/order";
import {FETCHING_STATUS} from "constants/api";

export const ReturnDeliveryConfirmationPopup = () => {
  const dispatch = useAppDispatch();
  const {
    data: {onDemandDeliveryStore, ownDeliveryStore, address},
  } = useSelector(getNearStoresData);
  const {pickup, selectedServices} = useSelector(getScheduleDetails);
  const {fetchingStatus: returnLoadingStatus} = useSelector(getReturnWindowsState);
  const [isLoadingReturn, setIsLoadingReturn] = useState(false);

  const closeView = () => {
    dispatch(onlineOrderActions.setStage(VIEWS.RECOMMENDED_PICKUP));
  };

  const turnAroundInHours =
    onDemandDeliveryStore?.turnAroundInHours || ownDeliveryStore?.turnAroundInHours;

  const onScheduleNowClick = () => {
    dispatch(
      onlineOrderThunks.getReturnWindows({
        windowsEndTime: pickup?.endTime,
        timeZone: pickup?.timeZone,
        storeId: ownDeliveryStore?.storeId || onDemandDeliveryStore?.storeId,
        zipCode: pickup?.zipCode,
        selectedServices,
        address,
      })
    );
  };

  const onScheduleLaterClick = () => {
    dispatch(onlineOrderActions.setStage(VIEWS.SUBSCRIPTION_OFFER));
  };

  useEffect(() => {
    if (returnLoadingStatus === FETCHING_STATUS.PENDING) {
      setIsLoadingReturn(true);
    }

    if (
      isLoadingReturn &&
      (returnLoadingStatus === FETCHING_STATUS.FULFILLED ||
        returnLoadingStatus === FETCHING_STATUS.REJECTED)
    ) {
      dispatch(onlineOrderActions.setStage(VIEWS.RECOMMENDED_RETURN));
      setIsLoadingReturn(false);
    }
  }, [dispatch, isLoadingReturn, returnLoadingStatus]);

  return (
    <div className="return-choice-container">
      <div className="close-container">
        <IconButton onClick={closeView}>
          <Close fontSize="large" />
        </IconButton>
      </div>
      <div className="title">
        <Typography variant="h1" component="h2">
          {SCHEDULE.returnDelivery}
        </Typography>
      </div>
      <div className="return-choice-content">
        <div className="choice-block">
          <div className="content">
            <Typography variant="h1" component="h2">
              {ORDER_MESSAGES.scheduleNow}
            </Typography>
            <img src={VanImg} alt="Van" />
          </div>
          <Button
            variant="contained"
            color="primary"
            onClick={onScheduleNowClick}
            className="schedule-button pop-up-button"
            startIcon={
              isLoadingReturn ? <CircularProgress color="inherit" size={14} /> : null
            }
          >
            {ORDER_MESSAGES.scheduleDeliveryNow}
          </Button>
        </div>
        <div className="choice-block">
          <div className="content">
            <Typography variant="h1" component="h2">
              {TEXT_ME_WHEN_READY} (
              {ORDER_MESSAGES.getTurnAroundInHours(turnAroundInHours)})
            </Typography>
            <img src={TurnAroundImg} alt="Turn Around" />
          </div>
          <Button
            variant="contained"
            color="primary"
            onClick={onScheduleLaterClick}
            className="schedule-button pop-up-button"
          >
            {ORDER_MESSAGES.scheduleDeliveryLater}
          </Button>
        </div>
      </div>
    </div>
  );
};
