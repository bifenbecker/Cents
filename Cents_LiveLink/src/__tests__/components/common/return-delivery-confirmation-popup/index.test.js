import React from "react";
import {render, fireEvent} from "@testing-library/react";

import ReturnDeliveryConfirmationPopup from "../../../../components/common/return-delivery-confirmation-popup/index.js";

const toggle = jest.fn();
const onScheduleLaterClick = jest.fn();
const turnAroundInHours = 12;
const onScheduleNowClick = jest.fn();

const renderReturnDeliveryConfirmationPopup = () => {
  const {container, queryByText, queryByAltText} = render(
    <ReturnDeliveryConfirmationPopup
      {...{
        isOpen: true,
        toggle,
        onScheduleLaterClick,
        turnAroundInHours,
        onScheduleNowClick,
      }}
    />
  );

  return {container, queryByText, queryByAltText};
};

describe("ReturnDeliveryConfirmationPopup Component", () => {
  it("should match snapshot", () => {
    const {container} = renderReturnDeliveryConfirmationPopup();
    expect(container).toMatchSnapshot();
  });

  it("should call toggle when back button is clicked", () => {
    const {queryByAltText} = renderReturnDeliveryConfirmationPopup();
    const backImg = queryByAltText("Dock Close");
    expect(backImg).toBeInTheDocument();

    fireEvent.click(backImg);
    expect(toggle).toBeCalled();
  });

  it("should have `Text me when it's ready` section", () => {
    const {queryByText, queryByAltText} = renderReturnDeliveryConfirmationPopup();

    expect(queryByText("Text me when it's ready (est. 12 hrs)")).toBeInTheDocument();
    expect(queryByAltText("Turnaround time without hand")).toBeInTheDocument();
    expect(queryByText("Schedule Delivery Later")).toBeInTheDocument();
  });

  it("should call onScheduleLaterClick function", () => {
    const {queryByText} = renderReturnDeliveryConfirmationPopup();

    const button = queryByText("Schedule Delivery Later");
    fireEvent.click(button);

    expect(onScheduleLaterClick).toBeCalled();
  });

  it("should have `I'll schedule my return delivery now` section", () => {
    const {queryByText, queryByAltText} = renderReturnDeliveryConfirmationPopup();

    expect(queryByText("I'll schedule my return delivery now")).toBeInTheDocument();
    expect(queryByAltText("Blue Van")).toBeInTheDocument();
    expect(queryByText("Schedule Delivery Now")).toBeInTheDocument();
  });

  it("should call onScheduleNowClick function", () => {
    const {queryByText} = renderReturnDeliveryConfirmationPopup();

    const button = queryByText("Schedule Delivery Now");
    fireEvent.click(button);

    expect(onScheduleNowClick).toBeCalled();
  });
});
