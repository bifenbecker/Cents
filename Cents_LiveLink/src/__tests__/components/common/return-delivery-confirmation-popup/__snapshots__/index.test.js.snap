// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReturnDeliveryConfirmationPopup Component should match snapshot 1`] = `
<div>
  <div
    style="position: fixed; width: 0px; height: 0px; top: 0px; left: 0px; z-index: 99999999;"
  >
    <div
      style="position: fixed; left: 0px; right: 0px; top: 0px; bottom: 0px; z-index: 0; opacity: 1; transition: opacity 0.2s ease-out; background: white;"
    />
    <div
      style="position: fixed; z-index: 1; box-shadow: 0 0 4px rgba(0, 0, 0, 0.3); background: white; left: 25%; top: 0px; width: 50%; height: 768px; transition: left 0.2s ease-out,top 0.2s ease-out,width 0.2s ease-out,height 0.2s ease-out;"
    >
      <div
        style="position: absolute; z-index: 2; opacity: 0; top: -5px; height: 10px; left: 0px; width: 100%; cursor: row-resize;"
      />
      <div
        style="width: 100%; height: 100%; overflow: auto;"
      >
        <div
          class="sc-bczRLJ sc-gsnTZi haujOk hzPtvw"
          font-size="18px"
          height="67px"
        >
          <img
            alt="Dock Close"
            class="sc-bczRLJ FYQvo"
            src="Icon_Back.svg"
          />
          <div
            class="sc-bczRLJ flhfNA"
            color="black"
            font-size="18px"
          >
            Return Delivery
          </div>
        </div>
        <div
          class="sc-bczRLJ kejCFv"
          display="grid"
        >
          <div
            class="sc-bczRLJ sc-gsnTZi ecSAXF hzPtvw"
            height="173px"
          >
            <div
              class="sc-bczRLJ sc-gsnTZi hUPHzA hzPtvw"
            >
              <div
                class="sc-bczRLJ IYlGL"
                font-size="20px"
              >
                I'll schedule my return delivery now
              </div>
              <img
                alt="Blue Van"
                class="sc-bczRLJ gMaQII"
                src="Blue_Van.svg"
                width="110px"
              />
            </div>
            <button
              class="sc-bczRLJ cGXRXt"
              font-family="primary"
              font-size="16px"
              width="100%"
            >
              Schedule Delivery Now
            </button>
          </div>
          <div
            class="sc-bczRLJ sc-gsnTZi ecSAXF hzPtvw"
            height="173px"
          >
            <div
              class="sc-bczRLJ sc-gsnTZi hUPHzA hzPtvw"
            >
              <div
                class="sc-bczRLJ IYlGL"
                font-size="20px"
              >
                Text me when it's ready
                 (est. 
                12
                 hrs)
              </div>
              <img
                alt="Turnaround time without hand"
                class="sc-bczRLJ gMaQII"
                src="TurnaroundTime_WithoutHand.svg"
                width="110px"
              />
            </div>
            <button
              class="sc-bczRLJ cGXRXt"
              font-family="primary"
              font-size="16px"
              width="100%"
            >
              Schedule Delivery Later
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
